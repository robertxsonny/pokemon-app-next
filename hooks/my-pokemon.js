import { useCallback, useEffect } from "react";
import { useState } from "react";

export const useMyPokemons = () => {
  const [caughtPokemons, setCaughtPokemons] = useState(() => {
    if (typeof window === "undefined") {
      return {}
    }

    const caughtPokemonsStr = localStorage.getItem('caught-pokemons');
    return caughtPokemonsStr ? JSON.parse(caughtPokemonsStr) : {};
  });

  useEffect(() => {
    localStorage.setItem('caught-pokemons', JSON.stringify(caughtPokemons));
  }, [caughtPokemons])

  const caughtPokemonNicknames = Object.keys(caughtPokemons);

  const hasPokemonWithNickname = useCallback((nickname) => caughtPokemonNicknames.includes(nickname), [caughtPokemonNicknames]);

  const getPokemonName = useCallback((nickname) => caughtPokemons[nickname], [caughtPokemons]);

  const getNumOfCollected = useCallback((name) => Object.values(caughtPokemons).filter(v => v === name).length, [caughtPokemons]);

  const catchPokemon = useCallback((name, nickname) => {
    if (!caughtPokemons[nickname]) {
      const newCaughtPokemons = {
        ...caughtPokemons,
        [nickname]: name
      }

      setCaughtPokemons(newCaughtPokemons);
    }
  }, [caughtPokemons]);

  const releasePokemon = useCallback((nicknameToRelease) => {
    const newCaughtPokemons = Object.keys(caughtPokemons).reduce((pokemons, nickname) => {
      return nickname === nicknameToRelease ? pokemons : { ...pokemons, [nickname]: caughtPokemons[nickname] }
    }, {})

    setCaughtPokemons(newCaughtPokemons);
  }, [caughtPokemons]);

  return { caughtPokemons, hasPokemonWithNickname, getNumOfCollected, getPokemonName, catchPokemon, releasePokemon };
}
