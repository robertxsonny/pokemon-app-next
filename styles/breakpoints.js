const generateBreakpoint = (bp) => ({ max: `@media (max-width: ${bp - 1}px)`, min: `@media (min-width: ${bp}px)` })

export const xs = generateBreakpoint(576);
export const sm = generateBreakpoint(768);
export const md = generateBreakpoint(992);
export const lg = generateBreakpoint(1200)