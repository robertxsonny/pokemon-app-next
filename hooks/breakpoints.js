import { useLayoutEffect, useState } from 'react';
import { xs, sm, md, lg, xl } from '../styles/breakpoints';

export const useBreakpoints = (breakpoint, layoutDir) => {
  const [width, setWidth] = useState(undefined);

  useLayoutEffect(() => {
    const handleResize = () => {
      setWidth(window.innerWidth);
    }

    window.addEventListener('resize', handleResize);
    handleResize();
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  const breakpoints = { xs, sm, md, lg, xl }
  
  if (!breakpoints.hasOwnProperty(breakpoint)) {
    return false;
  }

  const { min, max } = breakpoints[breakpoint];

  if (layoutDir === 'up') { return width >= min; }
  if (layoutDir === 'down') { return width <= max; }
  return (width >= min && width <= max);
}