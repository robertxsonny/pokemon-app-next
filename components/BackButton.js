import styled from '@emotion/styled';
import { useRouter } from 'next/router';

import { useBreakpoints } from '../hooks/breakpoints';
import { TextButton } from '../styles/shared';

const LargeTextButton = styled(TextButton)(
  {
    fontSize: 20,
    fontWeight: 500,
    marginRight: 16
  },
  ({ fixed }) => (fixed ? {
    position: 'fixed',
    top: 12,
    left: 24,
    zIndex: 9
  } : {})
)

const BackButton = ({ fixed }) => {
  const { back } = useRouter();
  const isSmallScreen = useBreakpoints('sm', 'down');

  return isSmallScreen ? <LargeTextButton fixed={fixed} onClick={back}>&lt;</LargeTextButton> : null;
}

export default BackButton;
