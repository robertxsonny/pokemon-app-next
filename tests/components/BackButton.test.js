import { fireEvent, render, screen } from '@testing-library/react';

import BackButton from '../../components/BackButton';
import { useBreakpoints } from '../../hooks/breakpoints';

const mockBackFunction = jest.fn();

jest
  .mock('../../hooks/breakpoints', () => ({
    useBreakpoints: jest.fn()
  }))
  .mock('next/router', () => ({
  useRouter: () =>({
    back: mockBackFunction
  })
}))

describe('BackButton component', () => {
  test('Should render normal back button', () => {
    useBreakpoints.mockReturnValueOnce(true);
    const { container } = render(<BackButton />);
    expect(container).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button'));
    expect(mockBackFunction).toHaveBeenCalled();
  })

  test('Should render fixed back button', () => {
    useBreakpoints.mockReturnValueOnce(true);
    const { container } = render(<BackButton fixed />);
    expect(container).toMatchSnapshot();
  })

  test('Should not render back button on large', () => {
    useBreakpoints.mockReturnValueOnce(false);
    const { container } = render(<BackButton />);
    expect(container).toMatchSnapshot();
  })
})
