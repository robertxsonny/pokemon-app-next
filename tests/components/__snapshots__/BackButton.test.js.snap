// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton component Should not render back button on large 1`] = `<div />`;

exports[`BackButton component Should render fixed back button 1`] = `
<div>
  <button
    class="css-1iz9xu8-TextButton-LargeTextButton e12p61ez0"
  >
    &lt;
  </button>
</div>
`;

exports[`BackButton component Should render normal back button 1`] = `
<div>
  <button
    class="css-1a6de82-TextButton-LargeTextButton e12p61ez0"
  >
    &lt;
  </button>
</div>
`;
